contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,572 @@
+
+BACKGROUND:
+
+For example, an admin has already created the devices locally and wishes to expose them to the user in a consistent manner through the 
+Persistent Volume API.
+
+WORKFLOW:
+
+USER:
+
+```
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: local-raw-pvc
+spec:
+  volumeMode: Block
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 80Gi
+```
+
+## UC4: 
+
+DESCRIPTION: StorageClass with non-dynamically created volumes
+
+BACKGROUND: The admin wishes to create a storage class that will identify pre-provisioned block PVs based on a user's PVC request for volumeMode: Block. 
+
+WORKFLOW: 
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: block-volume
+provisioner: no-provisioning 
+parameters:
+```
+* Sample of pre-created volume definition:
+
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+ name: pv-block-volume
+spec:
+ volumeMode: Block
+ storageClassName: block-volume
+ capacity:
+   storage: 35Gi
+ accessModes:
+   - ReadWriteOnce
+ local:
+    path: /dev/xvdc
+```
+## [FUTURE] UC5: 
+
+DESCRIPTION: StorageClass with dynamically created volumes 
+
+BACKGROUND: The admin wishes to create a storage class that will dynamically create block PVs based on a user's PVC request for volumeMode: Block. The admin desires the volumes be created dynamically and deleted when the PV definition is deleted. 
+
+WORKFLOW:
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: local-fast
+provisioner: kubernetes.io/local-block-ssd
+parameters:
+  fstype: Block #suggested value
+```
+
+***This has implementation details that have yet to be determined. It is included in this proposal for completeness of design ****
+
+## UC6: 
+
+DESCRIPTION: The developer wishes to request a block device via a Storage Class.
+
+WORKFLOW:
+
+USER: 
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+ name: pvc-local-block
+spec:
+ volumeMode: Block
+ storageClassName: local-fast
+ accessModes:
+  - ReadWriteOnce
+ resources:
+   requests:
+     storage: 10Gi
+```  
+
+## UC7: 
+
+DESCRIPTION: Admin creates network raw block devices
+
+BACKGROUND: Admin wishes to pre-create Persistent Volumes in GCE as raw block devices
+
+WORKFLOW:
+
+ADMIN:
+
+```
+apiVersion: "v1"
+kind: "PersistentVolume"
+metadata:
+  name: gce-disk-1
+  annotations:
+    volume.beta.kubernetes.io/mount-options: "discard"
+Spec:
+  volumeMode: Block
+  capacity:
+    storage: "10Gi"
+  accessModes:
+    - "ReadWriteOnce"
+  gcePersistentDisk:
+    fsType: "Block" #this is a suggestion, it is plugin dependent
+    pdName: "gce-disk-1"
+```
+***fsType values will be provisioner dependent. Block is suggested for development simplicity. Since the PVC object is passed
```

Comment:
Sorry, I really don't like special casing "Block" for fsType.  It's not a type of filesystem so doesn't match the meaning of the field.  Can we default to ext4 if PVC.volumeMode == filesystem for backwards compatability, and if PV.volumeMode == block, leave it empty?

Reply:
Changing a default value for a field in an existing API based on the presence/value of a new field not all clients (including version-skewed kubelets) know about is a real stretch, compatibility-wise

Reply:
If the client doesn't know about the new field, then they shouldn't be allowed to use the new feature, right?

Reply:
If this makes backwards compatibility/validation/defaulting too complicated, I could warm up to fsType = None.  I just think as a user, it's strange to need to set two fields correctly when the second field has no meaning in the case of block.

Reply:
@liggitt @msau42 I really don't think I should specify ANYTHING for the plugins. Or provide any example referring to fstype. We aren't mandating changes with it and it's not a 1st class field. Therefore, I think I should just remove the reference UC with it entirely with the suggestion that PVC be the driver for the request and that plugins should respect the user request as best as possible. This is the common decisioning we already do in binding.

Reply:
@liggitt @msau42 I will leave in the section on Storage Classes for completeness of E2E design and future pieces but I have removed the UC with it directly called out

Reply:
Sounds good, it'll be up to each plugin to decide how to handle it.